packages/server/src/api/controllers/query/index.ts

```diff
@@ -142,6 +142,78 @@ export async function preview(ctx: UserCtx) {
 
   const authConfigCtx: any = getAuthConfig(ctx)
 
+  function getFieldMetadata(
+    field: any,
+    key: string,
+    nestedSchemaFields: {
+      [key: string]: Record<string, string | QuerySchema>
+    }
+  ): QuerySchema {
+    const makeQuerySchema = (
+      type: FieldType,
+      name: string,
+      subtype?: string
+    ): QuerySchema => ({
+      type,
+      name,
+      subtype,
+    })
+    let type = typeof field,
+      fieldMetadata = makeQuerySchema(FieldType.STRING, key)
+    if (field)
+      switch (type) {
+        case "boolean":
+          fieldMetadata = makeQuerySchema(FieldType.BOOLEAN, key)
+          break
+        case "object":
+          if (field instanceof Date) {
+            fieldMetadata = makeQuerySchema(FieldType.DATETIME, key)
+          } else if (Array.isArray(field)) {
+            const firstFiveItems = field.slice(0, 5)
+            if (firstFiveItems.some(item => JsonUtils.hasSchema(item))) {
+              fieldMetadata = makeQuerySchema(
+                FieldType.JSON,
+                key,
+                JsonFieldSubType.ARRAY
+              )
+            } else {
+              fieldMetadata = makeQuerySchema(FieldType.ARRAY, key)
+            }
+            nestedSchemaFields[key] = buildSchemaFromArray(
+              firstFiveItems,
+              nestedSchemaFields
+            )
```

Comment:
This modification as a side effect in a function called `getFieldMetadata` doesn't feel great, but it would likely require a substantial refactor to unwind. No action required, just felt compelled to point this out. If you can think of some way to make this nicer that doesn't completely derail the PR, feel free, but don't worry if you can't.

Reply:
Good call - I've put this modification into its own function